File: src/tutorial/go2_tutorial/src/go2_state.cpp
/*
需求：订阅里程计消息，当机器人位移超出指定阀值，输出机器人此刻的坐标。
实现流程：
    1.订阅里程计消息；
    2.在订阅方的回调函数中，计算当前机器人位置与上一个记录点的距离，
        如果距离大于阀值就输出坐标，并更新记录点。
    第一个记录点如何生成？
    当第一次订阅到里程计消息时，就为记录点赋值。
*/


#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
 
class Go2State: public rclcpp::Node{
public:
    Go2State(const std::string &node_name):Node(node_name){
        RCLCPP_INFO(this->get_logger(), "Go2State 创建");

        is_first = true;
        last_x = last_y = 0.0;

        odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>("odom", 10,
            std::bind(&Go2State::odom_cb, this, std::placeholders::_1)
        );

        this->declare_parameter("distance", 0.5);
    }
private:
    void odom_cb(const nav_msgs::msg::Odometry::SharedPtr odom){
        // 初始化第一个记录点数据
        // 获取当前坐标
        double x = odom->pose.pose.position.x;
        double y = odom->pose.pose.position.y;

        // 如果是第一次订阅到该消息，那么就赋值
        if(is_first){
            last_x = x;
            last_y = y;
            is_first = false;
            RCLCPP_INFO(this->get_logger(), "起点坐标：(%.2f, %.2f)", last_x, last_y);
            return;
        }

        //计算是否超出阀值
        double distance_x = x - last_x;
        double distance_y = y - last_y;
        // 计算距离
        double distance = sqrt(pow(distance_x,2) + pow(distance_y,2));
        auto dis = this->get_parameter("distance").as_double();
        if(distance > dis){
            RCLCPP_INFO(this->get_logger(), "坐标：(%.2f, %.2f)", x, y);
            // 更新记录点
            last_x = x;
            last_y = y;
        }
    }

    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
    double last_x, last_y;
    bool is_first;
};
 
int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<Go2State>("go2_state");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/tutorial/go2_tutorial_py/go2_tutorial_py/go2_state.py
"""
需求：
订阅里程计消息，当机器人位移超出指定阀值，输出机器人此刻的坐标。

实现流程：
    1. 订阅里程计消息；
    2. 在回调函数中，计算当前机器人位置与上一个记录点的距离；
        如果距离大于阀值，就输出坐标，并更新记录点；
    3. 第一个记录点在第一次接收到里程计消息时生成。
"""

import math

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry


class Go2StatePy(Node):

    def __init__(self):
        super().__init__('go2_state_py')
        self.get_logger().info('Go2StatePy 创建')

        # 是否第一次接收到里程计
        self.is_first = True
        self.last_x = 0.0
        self.last_y = 0.0

        # 订阅里程计
        self.odom_sub = self.create_subscription(Odometry, '/odom', self.odom_cb, 10)
        
        # 声明参数：距离阀值
        self.declare_parameter('distance', 0.5)
        
    def odom_cb(self, odom: Odometry):
        # 获取当前坐标
        x = odom.pose.pose.position.x
        y = odom.pose.pose.position.y

        # 第一次接收到消息，初始化记录点
        if self.is_first:
            self.last_x = x
            self.last_y = y
            self.is_first = False
            self.get_logger().info(f'起点坐标：({self.last_x:.2f}, {self.last_y:.2f})')
            return

        # 计算位移
        dis_x = x - self.last_x
        dis_y = y - self.last_y
        distance = math.sqrt(dis_x ** 2 + dis_y ** 2)

        # 获取阀值参数
        # threshold = self.get_parameter('distance').value  # 推荐
        threshold = self.get_parameter('distance').get_parameter_value().double_value
        # 判断是否超过阀值
        if distance > threshold:
            self.get_logger().info(f'坐标：({x:.2f}, {y:.2f})')
            # 更新记录点
            self.last_x = x
            self.last_y = y


def main(args=None):
    rclpy.init(args=args)
    node = Go2StatePy()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
    